### ISSUE TITLE
IsVisible false is not propagated to child controls

### ISSUE THREAD

**Issue by omghb** [dotnet/maui#19139]
### Description

If `IsVisible="False"` is set on a `Grid` then the child controls still return `true` for `IsVisible`.

On iOS this creates a bug as the keyboard "Next" button focus the invisible `Entry` an let the user type text into it.

### Steps to Reproduce

1. Create a new MAUI 8 app
2. Open `MainPage.xaml.cs` and replace the content with:
```C#
namespace MauiApp1
{
    public partial class MainPage : ContentPage
    {
        public MainPage()
        {
            InitializeComponent();
        }

        private void HiddenEntry2Focused(object sender, FocusEventArgs e)
        {
            Console.WriteLine("Hidden entry: IsVisible: " + hiddenEntry2.IsVisible);
        }

        private void Entry3Focused(object sender, FocusEventArgs e)
        {
            Console.WriteLine("Hidden entry: IsVisible: " + hiddenEntry2.IsVisible);
        }
    }
}
```
4. Open `MainPage.xaml` and replace the content with
```xaml
<?xml version="1.0" encoding="utf-8" ?>
<ContentPage xmlns="http://schemas.microsoft.com/dotnet/2021/maui"
             xmlns:x="http://schemas.microsoft.com/winfx/2009/xaml"
             x:Class="MauiApp1.MainPage">
    <ScrollView>
        <VerticalStackLayout Padding="30,0" Spacing="25">
            <Entry Placeholder="entry1" ReturnType="Next"/>

            <Grid IsVisible="False">
                <Entry x:Name="hiddenEntry2" Placeholder="hiddenEntry2" ReturnType="Next" Focused="HiddenEntry2Focused"/>
            </Grid>

            <Entry Placeholder="entry3" ReturnType="Next" Focused="Entry3Focused"/>
        </VerticalStackLayout>
    </ScrollView>
</ContentPage>
```
5. Run the App on iOS
6. Tap into the first entry and tap on the next button.

Result: Keyboard is shown for the hidden entry.

![Maui8IosVisibilityIssue](https://github.com/dotnet/maui/assets/47421673/246e94f3-c717-49f9-9c89-a9d7d64a153e)

### Version with bug

8.0.3

### Is this a regression from previous behavior?

Not sure, did not test other versions

### Last version that worked well

Unknown/Other

### Affected platforms

iOS, Android

**Comment by mattleibow** [comment]
I tried this and was not able to reproduce this.  [Archive.zip](https://github.com/dotnet/maui/files/13516544/Archive.zip)

**Comment by mattleibow** [comment]
Ah, it happens on iOS when using the keyboard "next" button 😱

**Comment by omghb** [comment]
@mattleibow The keyboard "next" button is working on Android as expected but the code line
```
Console.WriteLine("Hidden entry: IsVisible: " + hiddenEntry2.IsVisible);
```
returns `true` on Android as well. This should be `false` as the parent Grid has set `IsVisible="False"`.

**Comment by cucumber-sp** [comment]
@omghb no, it shouldn't be false. IsVisible on parent stops its rendering and rendering of children, but it doesn't change their properties.

**Comment by omghb** [comment]
> @omghb no, it shouldn't be false. IsVisible on parent stops its rendering and rendering of children, but it doesn't change their properties.

@cucumber-sp That's not what I would expect. How would I find out (via code) that a Element is not visible (not rendered)?

**Comment by tj-devel709** [comment]
Yeah this behavior for the next button seems incorrect, I will take a look at this, thanks!

**Comment by bradencohen** [comment]
> @omghb not sure if there's a built-in solution for this, but I guess you can just go up untile parent is null, if some control has IsVisible set to false return false otherwise true. Make it and extension function and that's it

@omghb I'm curious what the `Visibility` property would return. 

For ex: `var visibility = ( (IView) Content ).Visibility`

https://learn.microsoft.com/en-us/dotnet/api/microsoft.maui.iview.visibility?view=net-maui-8.0#microsoft-maui-iview-visibility

**Comment by omghb** [comment]
@bradencohen The `Visibility` property returns the same value as `IsVisible`. The element is hidden and should be `Collapsed`.

![image](https://github.com/dotnet/maui/assets/47421673/ac35c7cb-ab19-4b76-91c8-2ba0e8574c77)

**Comment by kubaflo** [comment]
@jsuarezruiz @mattleibow @omghb @omghb  PR: https://github.com/dotnet/maui/pull/20154

**Comment by mattleibow** [comment]
I just tested WPF and WinUI 3 and they do not propagate the value. We don't _have_ to follow the WPF way, but will the difference be confusing?

Anyone have any thoughts?

**Comment by mattleibow** [comment]
> Note that WPF has Visibility and IsVisible which are fundamentally different. The former represents the intent for that particular element, the latter is read-only, calculated visibility, which I guess is "propagated" in the sense you are thinking of (i.e. you cannot have IsVisible element until all parents are visible). Does that help?

Seems that this is actually a thing, so I am thinking maybe we should dio this now?

**Comment by omghb** [comment]
Thanks for the improvement in MAUI 8.0.10: #20160 

Unfortunately, this fixes the issue only partly. Still not working:
- Hide an element that has the focus
  - Expected: next focusable element gets the focus
  - Actual: focus stays on the hidden element
- IsVisible property is not propagated to child controls

**Comment by jfversluis** [comment]
We initially merged https://github.com/dotnet/maui/pull/20154 as a fix, but this caused some regressions so unfortunately we had to revert that one. With that, this still isn't completely fixed so reopening.

**Comment by mattleibow** [comment]
@bmacombe said

> > Now the question is... Is this a breaking change for a SR that means it should wait for net9.0? Or is it OK and people need to be aware of the change?
> 
> @mattleibow @kubaflo I think this is causing me issues in 9.0.50, we have some layout logic that depends on knowing the actually Set value of IsVisible, like the original behavior. I don't see how I can get the _isVisibleExplicit value even after the .net 10 changes

Are you doing layout on a control that is hidden?

**Comment by bmacombe** [comment]
@mattleibow My use case is actually pretty simple. I have a Grid that I dynamically add star width columns to based on the child count. Just to evenly space the controls across the screen.  But I'm only showing (and adding columns for) Visible controls. And if none of the controls are visible, I make the grid not visible until one of its child controls is visible again.

I mainly use it in data input views that show or hide input controls based on the user rights, etc.

**Comment by kubaflo** [comment]
Closing as completed because this issue is explicitly marked as shipped in `fixed-in-9.0.0-preview.2.10293`, `fixed-in-8.0.10`.

> Disclosure: this comment was generated by AI.

**Comment by omghb** [comment]
@kubaflo I tried this out with the latest MAUI version 10.0.90 - but it does not work. Here is what I checked again: https://github.com/dotnet/maui/issues/19139#issuecomment-1867642524
- Actual result: The `hiddenEntry2` has still the state visible.
- Expected result: The `hiddenEntry2` should have the state invisible or collapsed.

See also comment from @jfversluis: https://github.com/dotnet/maui/issues/19139#issuecomment-2789397678
> We initially merged [#20154](https://github.com/dotnet/maui/pull/20154) as a fix, but this caused some regressions so unfortunately we had to revert that one. With that, this still isn't completely fixed so reopening.
